Automattic/jetpack
# Build: Use PHPCompatibilityWP and more PHPCS related tweaks

## Commit Summary:

### Build/PHPCS: rename ruleset

If the PHPCS ruleset is named `.phpcs.xml`, `phpcs.xml`, `.phpcs.xml.dist` or `phpcs.xml.dist`, it will automatically be picked up by PHPCS and you don't need to pass the ruleset name anymore.

Additionally, using a `.dist` file for a repo ruleset allows for individual developers to overrule the ruleset with a custom version (without the `.dist` file extension).
This makes testing of new additions/changes to the ruleset easier.
The master ruleset can be imported into a custom ruleset by using `<rule ref="./.phpcs.xml.dist"/>`.

Since PHPCS 3.1.0, it is also possible to use dot-prefixed files for the PHPCS config, allowing these files to be sorted with other configuration related files.

The loading order of the ruleset files in PHPCS, as of version 3.1.1., is:
1. `.phpcs.xml`
2. `phpcs.xml`
3. `.phpcs.xml.dist`
4. `phpcs.xml.dist`

References:
* https://github.com/squizlabs/PHP_CodeSniffer/wiki/Advanced-Usage#using-a-default-configuration-file
* https://github.com/squizlabs/PHP_CodeSniffer/releases/tag/3.1.0
* https://github.com/squizlabs/PHP_CodeSniffer/releases/tag/3.1.1 (change in the file loading order)

### Build/PHPCS: Use PHPCompatibilityWP

This PR:
* Switches the repo over to use `PHPCompatibilityWP` rather than `PHPCompatibility`.
    As this repo and ruleset is aimed at a WordPress project, using the `PHPCompatibilityWP` ruleset is the better choice to prevent false positives for PHP functions and constants which are back-filled by WP.
* Switches the dependency over to use the repo in the `PHPCompatibility` organisation rather than the one in `wimg`'s personal account.
* Fixes other references to the PHPCompatibility standard.

#### Notes:
* Other than for the changed dependency, no other Composer updates have been run.

#### References:
* https://github.com/PHPCompatibility/PHPCompatibilityWP
* PHPCompatibility/PHPCompatibility#688
* https://github.com/PHPCompatibility/PHPCompatibility/releases/tag/8.2.0

### PHPCS: remove outdated PHPCS annotations

Now the repo uses `PHPCompatibilityWP` instead of `PHPCompatibility`, a number of `// phpcs:ignore` annotations can be removed.

### PHPCS: remove superfluous PHPCS annotations

The annotation in the `modules/custom-css/csstidy/class.csstidy_print.php` file wasn't ignoring anything.
The annotation in the `modules/custom-css/csstidy/class.csstidy.php` file was hiding some very real issues with the code.

While I understand that the code is sourced elsewhere and should be fixed at the source and this file synced after that, shipping the code as is without applying the suggested fixes is not a good idea either.

### PHPCS: improve maintainability of PHPCS annotations in WP-CLI file

The `WP-CLI` file has a minimum PHP version of 5.3 - in line with WP-CLI itself.
So, I've changed the individual inline annotations to a file-based annotation for easier maintenance.

### PHPCS: make remaining PHPCS annotations more specific

For the remaining ignore annotations, those have been verified and where applicable, made more specific to only ignore the issue(s) currently being reported and prevent issues which may be detected by future sniffs from *not* being reported.

### PHPCS: whitelist a couple of correct uses of new PHP functionality

When wrapped in a `function_exists()`, `class_exists()`, `defined()` etc, new PHP functionality can be safely used.

## Current PHP cross-version compatibility status:

After this PR has been merged, the current codebase (still) contains the following - verified - issues:

```
PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
------------------------------------------------------------------------------------------
STANDARD  CATEGORY            SNIFF                                                  COUNT
------------------------------------------------------------------------------------------
PHPCompa  PHP                 Discouraged switch continue found                      9
PHPCompa  PHP                 New function parameters debug_backtrace_options found  4
PHPCompa  PHP                 New closure found                                      3
PHPCompa  PHP                 Deprecated functions each deprecated                   2
PHPCompa  PHP                 New function parameters getenv_local_only found        2
PHPCompa  PHP                 Argument functions usage in parameter list             1
PHPCompa  PHP                 New function parameters json_encode_options found      1
------------------------------------------------------------------------------------------
A TOTAL OF 22 SNIFF VIOLATIONS WERE FOUND IN 7 SOURCES
------------------------------------------------------------------------------------------
```


## Testing instructions:

* Run `composer php:5.2-compatibility` to see the improved results of the PHPCompatibility scan.
